File: day_2.ts
import { readFileSync } from 'fs';
import { join as joinPath } from 'path';

const input: string = readFileSync(joinPath(__dirname, './input/2.txt'), 'utf-8');

interface Game {
  id: number;
  red: number;
  green: number;
  blue: number;
}

function parseGame(input: string): Game {
  const [id, draws] = input.split(': ');

  const game: Game = { id: parseInt(id.split(' ')[1]), red: 0, green: 0, blue: 0 };

  draws.split(/[,;] /).map(draw => {
    const [quantity, colour] = draw.split(' ');
    game[colour] = Math.max(game[colour], parseInt(quantity));
  });

  return game;
}

function parseInput(input: string): Game[] {
  return input.split('\n').map(line => parseGame(line));
}

function isPossible({ red, green, blue }: Game): boolean {
  return red <= 12 && green <= 13 && blue <= 14;
}

function gamePower({ red, green, blue }: Game): number {
  return red * green * blue;
}

const games: Game[] = parseInput(input);

console.log('Part 1:');
console.log(
  games.filter(isPossible).reduce((acc, { id }) => (acc + id), 0)
);

console.log('Part 2:');
console.log(
  games.map(gamePower).reduce((acc, power) => (acc + power), 0)
);

File: day_2.rb
#!/usr/bin/env ruby

COLOURS = %w[red green blue]

UPPER_BOUNDS = {
  red:   12,
  green: 13,
  blue:  14
}

class Game
  def initialize(line)
    id, hands = line.split(': ')
    @id = id.split(' ').last.to_i
    @hands = hands.split('; ').map { |hand| Hand.new hand }
  end

  attr_reader :id, :hands

  def max(colour)
    hands.map(&colour).max
  end

  def in_bounds?
    UPPER_BOUNDS.all? { |colour, limit| max(colour) <= limit }
  end

  def power
    max(:red) * max(:green) * max(:blue)
  end
end

class Hand
  def initialize(hand)
    @red = @green = @blue = 0

    hand.split(', ').each do |draw|
      amount, colour = draw.split(' ')

      send("#{colour}=", amount.to_i)
    end
  end

  attr_accessor :red, :green, :blue
end

TEST_GAMES = File.readlines("./input/2.test").map { |line| Game.new(line) }
GAMES = File.readlines("./input/2.txt").map { |line| Game.new(line) }

def part_one(games)
  games.select(&:in_bounds?).map(&:id).sum
end

def part_two(games)
  games.map(&:power).sum
end

puts 'PART 1:'
puts part_one TEST_GAMES
puts part_one GAMES
puts 'PART 2:'
puts part_two TEST_GAMES
puts part_two GAMES
